### ISSUE TITLE
[Feature] add `--sparse` option to `get`, `pin add` subcommands, for broken file

### ISSUE THREAD

**Issue by elitak** [ipfs/go-ipfs#5188]
#### Description:
Sometimes I have a file in my local repo that's missing a single block, that's nowhere on the network either. When I `ipfs get <multihash>`, it stalls at the position where that block is in the file, forever. Media files usually tolerate a single missing block just fine, so I'd like a `--sparse` option that would simply skip unavailable blocks and leave it unwritten in a sparse file or nulled in filesystems that don't support sparse files. *Unavailable* would mean that the block stalls (0bps downstream) longer than X seconds or takes Y% (~1000%) times longer than the other blocks to retrieve. Maybe X and Y could be configurable in the daemon's settings.

Similarly, a `--sparse` or maybe `--partial` option to `ipfs pin add` would skip over the missing blocks and pin the rest, while still returning some non-zero error, because the entire file could not be pinned.

The other motivation for this besides playing media files is that when I lose pieces of a file and want to recover them from a torrent, I have no easy way to get the trailing good blocks out of the ipfs datastore. I presume the API allows me to fetch each block individually and write it into the file myself, but I think this facility should be part of the CLI. Likewise, when pinning a file, I sometimes want to transfer over as much as possible in the background, while working on recovering the last piece to complete the pin.

**Comment by schomatis** [comment]
Hey @elitak, thanks for raising this issue, it's a very interesting proposal.

To add some more context to it let me clarify that files are organized in DAGs (a tree, roughly speaking), so different nodes (stored in the blocks you mentioned) at different depths represent different parts of the file and there is not necessarily a linear relationship between the node size (block size) and the file size it represents, meaning that nodes higher in the tree represent larger chunks in the file and if they get lost (unavailable as you say) a big part of the file gets lost (to illustrate this, in the more extreme case, if you lose the root you don't have a file). I'm saying this to contrast it with the usual streaming scenario in which you lose a single UDP packet, a small chunk of a file, and pretty much nothing happens (you won't notice it I mean).

So maybe this could also be taken in consideration, another threshold could be set with the maximum number of contiguous bytes that can be skipped without losing perceivable streaming quality.

**Comment by elitak** [comment]
Yes, thanks, I hadn't realized that. Still, my primary use case is to recover broken files as best as possible. Right now, if a block is missing 20% in, I can only easily dump that leading series of blocks onto disk, ctrl-c once it hangs, then verify the output using the original torrent, and re-download the other 80%. Usually, when these files are media and i play them using a http URI through the API, I'm able to seek through most of the file outside of that 1 problem area.

Before anybody asks, let me say that the reason I'm losing blocks in the first place is that I've been running the daemon on a very low-spec ARM host, and for whatever reason the LUKS+xfs and even LUKS+ext4 tends to sustain a fair amount of corruption over time (the physical disk is new). I'm pretty confident there's little the daemon could be doing differently to prevent this. That said, another nice feature would be a way to more quickly verify a pin, without rehashing, to at least check that the block files in the datastore are present and the expected size on disk. I may file this later.

**Comment by schomatis** [comment]
To help you get started here's a simple script that removes a block in the middle of an added string.

```
# Split each char in its own block.
echo "0123X56789" | ipfs add --chunker=size-1 --raw-leaves -Q
# QmQJj1Fu6cFiKFxwbTFSDZ7jMMQqoKzGHiDaURMtv1FiH3

ipfs cat QmQJj1Fu6cFiKFxwbTFSDZ7jMMQqoKzGHiDaURMtv1FiH3
# 0123X56789

# Remove the block that contains the 'X' char.
rm ~/.ipfs/blocks/AC/AFKREICLNCVTQR763J6WYYWB7PF65P5DL2VXGUPNLZ4PJXNN5JO7MS4ACU.data 

ipfs cat QmQJj1Fu6cFiKFxwbTFSDZ7jMMQqoKzGHiDaURMtv1FiH3
# 0123Error: failed to fetch all nodes
# (There's a newline missing.)

```

**Comment by elitak** [comment]
Yes, your example captures the basic idea. Building off that script:
```
$ dd of=expected if=<(echo -n 0123) bs=1
$ dd of=expected if=<(echo -n 56789) bs=1 seek=5
$ ls -s expected                                                                                                       
4 expected
$ hexdump -C expected                                                                                                   
00000000  30 31 32 33 00 35 36 37  38 39                    |0123.56789|
0000000a
$ diff -q <(ipfs cat --sparse QmQJj1Fu6cFiKFxwbTFSDZ7jMMQqoKzGHiDaURMtv1FiH3) expected # should be equal, as long as that block cannot be retrieved
```

In this case, the file is not truly sparse, because the block size is not a multiple of the inode size for the filesystem, hence no space savings occur. To write truly sparse files would be more of a non-functional requirement, anyway; it's likely to be of no consequence, other than saving a tiny bit of disk space.

**Comment by schomatis** [comment]
Ok, this could be implemented from the DAG reader (I think), if a child node is missing (the promise is not fulfilled) the reader could output a number of null bytes (or similar) equal to the `Blocksize` (child size) indicated by the parent node and move on (when this option is set). Yes, this wouldn't be a "sparse" option, rather an "allow-missing-blocks" option.

@whyrusleeping WDYT?

**Comment by Stebalien** [comment]
I'm not sure this really needs a new external-facing feature. We can probably implement this the same way bittorrent clients do: write the chunks as we receive them, always keep looking for new chunks. That is, you can call `ipfs get ...` on a file and it will:

1. Allocate the file up-front.
2. Keep fetching other parts of the file even it it gets stuck fetching some parts.

Now, keeping the wantlist manageable could be tricky but we need to work on that anyways.

---

As for pin, you can:

1. Add the file to MFS.
2. Call `ipfs refs /path/to/file` to fetch what you can.

However, like `get`, that command will hang if it encounters a block it can't find. That'll need to be fixed. Also, it would be nice to have a way to tell MFS to continue downloading data in the background (but that's a separate feature I've wanted for a while).
